### ISSUE TITLE
training gets stuck with sync bn

### ISSUE THREAD

**Issue by jiang1st** [tensorpack/tensorpack#1078]
Hi, I am replacing standard batch norm with sync batch norm in my network. I am using 8 P100 on centos 7.0.  I have installed nccl 2.4.2 + cuda9.0. The backbone is mobilenet v2. Here is my problem:

When I set:   `with argscope(BatchNorm, momentum=0.997, epsilon=0.001, training=True, sync_statistics='nccl')` , after building the graph, the training phase gets stuck at:

  0%|                                                                                                                              |0/300[00:00<?,?it/s][0212 15:59:37 @input_source.py:551] Pre-filling StagingArea ...
[0212 15:59:39 @input_source.py:555] 1 element was put into StagingArea on each tower.

But when I set with `argscope(BatchNorm, momentum=0.997, epsilon=0.001, training=True)` , then the training phase goes successfully.

I guess there is something wrong with my sync batch norm. What could be the reason? Thank you.

**Comment by ppwwyyxx** [comment]
https://tensorpack.readthedocs.io/modules/models.html#tensorpack.models.BatchNorm has mentioned the common reasons why it may hang.

If you cannot figure out the reason, please try running the built-in example with syncbn first -- if that fails, please fill the issue template.

**Comment by jiang1st** [comment]
I tried one built-in example with sync bn, and it works fine. It should be my problem when replacing slim.batch_norm with tensorpack's batch norm. Will check it.

**Comment by ppwwyyxx** [comment]
Another possibility:

In every iteration, either all GPUs execute the BatchNorm once, or no GPUs execute the BatchNorm. If only a subset of GPUs execute the BatchNorm, or if the BatchNorm is executed more than once, this layer may hang because there is nothing to sync.

**Comment by jiang1st** [comment]
Thank you for your suggestion. I think the operations in each gpu are the same. I used SyncMultiGPUTrainerReplicated from tensorpack for multi-gpu training.

When would the BatchNorm be executed more than once? I tried to print the tensor name of the input at the beginning of BatchNorm, it looks like each BatchNorm is called only once.

**Comment by ppwwyyxx** [comment]
> When would the BatchNorm be executed more than once?

For example when it is executed in a while loop

**Comment by ppwwyyxx** [comment]
You may also want to remove some custom callbacks to see if they affect this. Callbacks in general should not use `hooked_sess` and it may cause issues like this if you use `hooked_sess`.

**Comment by jiang1st** [comment]
Thank you for the suggestions. I think I have put the batch_norm in a `for` loop ([ slim mobilenet v2](https://github.com/tensorflow/models/blob/master/research/slim/nets/mobilenet/mobilenet.py#L248) ). If a put batch_norm outside the `for` loop, then it works well.

What I don't understand is that, in the `for` loop, layers all have different names (including batch_norm). Why would batch_norm hang? Thank you.

**Comment by ppwwyyxx** [comment]
> I have put the batch_norm in a `for` loop ([ slim mobilenet v2](https://github.com/tensorflow/models/blob/master/research/slim/nets/mobilenet/mobilenet.py#L248) )

This is unrelated. What I meant is a symbolic loop (like `tf.while_loop`) which will execute the same layer in each iteration.
A loop in python side like this will create a new layer in each iteration.

**Comment by ppwwyyxx** [comment]
> If a put batch_norm outside the `for` loop, then it works well.

That sounds like you may have used the same name for different batchnorm layers.

**Comment by ppwwyyxx** [comment]
You can `print(shared_name)` at https://github.com/tensorpack/tensorpack/blob/49675590da8a39c649b5f0f5ba522a22b90e2d69/tensorpack/models/batch_norm.py#L248
and see if there are any duplications within one GPU, and if different GPUs produce the same name.

**Comment by jiang1st** [comment]
Thank you for the suggestions. I think I have found the reason: I only used the first few layers of mobilenet backbone for training. The last several layers are not used for training, but were created by default when loading `mobilenet_base()`. After removing the batch_norm in these layers, the training works.

**Comment by ppwwyyxx** [comment]
Thanks a lot for finding this out! 
Indeed, unused batchnorm will also cause this. I'll fix this or add this note to documentation.

**Comment by ppwwyyxx** [comment]
Btw, it will not cause this issue if you use the `internal_update=True` option in `BatchNorm`.

**Comment by ppwwyyxx** [comment]
Should've been fixed after https://github.com/tensorpack/tensorpack/commit/c366eebf04f5e53022b1b266cc0f940bf6789f61
Docs (https://tensorpack.readthedocs.io/modules/models.html#tensorpack.models.BatchNorm) was also updated.

In general, the model will run faster with `internal_update=True` when some batchnorm are unused. This was also now mentioned in the docs.
The default was kept as `internal_update=False` to be consistent with `tf.layers`.

**Comment by jiang1st** [comment]
Thanks for the update!
